forcedotcom/sfdx-core
# feat!: cross-save resistant configFile

> requires https://github.com/salesforcecli/plugin-data/pull/721

### Primary: make configFile save from cross-saving
- every property in a file is now a Last-Write-Wins Register on a Conflict Free Replicated Data Type, and BaseConfigStore now holds the CRDT
  - they record timestamps, and the more recent timestamp wins
- when a ConfigFile is read from fs, the fs modstamp is used as the timestamp for every property in the CRDT
- deletes (removed properties) are handled with a special tombstone in their state
- lockfiles (same pattern uses to fix the `alias` cross-save problem) are used in the write/writeSync methods

BREAKING CHANGE: `AuthInfo.getFields` now returns a read-only object.  Use AuthInfo.update to change values in the fields.

BREAKING CHANGE: `setContents` method is no longer available in the ConfigFile stack.  
BREAKING CHANGE:`awaitEach` is removed from ConfigFile stack
BREAKING CHANGE: `write(sync)` method no longer accepts a param.  Use other methods (set, unset) to make modifications, then call `write()`/`writeSync()` to do the write.  
BREAKING CHANGE: the use of lodash-style get/set (ex: `set('foo.bar.baz[0]', 3)` no longer works.
BREAKING CHANGE: You can no longer override the `setMethod` and `getMethod` when extending classes built on ConfigFile.  Technically you could override get/set, but DON'T!
BREAKING CHANGE: everything related to tokens/tokenConfig is gone.

### Other changes

- BREAKING CHANGE: node18+ only, compiles to es2022
- move uniqid to a shared function, outside of testSetup

### Remove previously deprecated stuff since this is a major release

- BREAKING CHANGE: removed `sfdc.isInternalUrl`.  Use `new SfdcUrl(url).isInternalUrl()`
- BREAKING CHANGE: removed `sfdc.findUppercaseKeys`.  There is no replacement.
- BREAKING CHANGE: removed `SchemaPrinter`.  There is no replacement.

also `ttypescript` => `ts-patch` [@W-14388558@](https://gus.lightning.force.com/a07EE00001dQzJ5YAK)

TODO:

- [x] make sure this publishes a major version
- [x] setContents, especially with `sfdxConfig.merge`
- [x] plugin-data adjustments
- [x] what to do about getMethod/setMethod
- [x] can we do without the CRDT peer/id stuff?
- [x] local NUTs for massive parallelization of stuff hitting the same file
- [x] migration docs
- [x] check command perf in plugin-settings

### What issues does this PR fix or reference?
https://github.com/forcedotcom/cli/issues/2423
[@W-14085765@](https://gus.lightning.force.com/a07EE00001a9pRDYAY)

## QA highlights

The code for `sfdxConfig` interop changed quite a bit.  Make sure it's still reading/writing that file correctly.
Compare perf in things like plugin-settings and plugin-auth where we're writing to these files.  the lock stuff could introduce some extra latency.

Check cases (ex: no file exists)

Things outside of core that use ConfigFile to test with
- deployCache and manifestCache in PDR
- scratch and sandbox caches in plugin-org
- something in plugin-data for supporting --use-most-recent
- remoteSourceTrackingService in STL
- most of plugin-settings (config and alias)
